README.md
# openai_ros

## Capabilities
 
This **ROS Noetic** package provides similiar functions to the [panda-gym](https://github.com/qgallouedec/panda-gym) package, but based on Gazebo. Here, we implemented several robotics environments to allow using this package to perform Reinforcement Learning on tasks, specifically for the **Franka Emika Robot**. It is directly modified from , and follows similar structure with the [openai_ros](https://bitbucket.org/theconstructcore/openai_ros/src/kinetic-devel/) package by TheConstructCore.

## Installation

Execute the following commands:<br>
`cd ~/ros_ws/src`<br>
`git clone https://github.com/angiecodinghub/openai_ros.git src/openai_ros`<br>
`cd ~/ros_ws`<br>
`catkin_make`<br>
`source devel/setup.bash`<br>
`rosdep install openai_ros`<br>

## Available Task Environments

1. Reach: <br>
The goal is to let the end effector of the robot reach a certain location in the space. 
* gripper: blocked. 
* maximum step per episode: 50.
* init position: 90 degrees at its elbow: <br>
            {'panda_joint1': 0.0, <br>
            'panda_joint2': 0.0, <br>
            'panda_joint3': 0.0, <br>
            'panda_joint4': -1.57079632679, <br>
            'panda_joint5': 0.0, <br>
            'panda_joint6': 1.57079632679, <br>
            'panda_joint7': 0.785398163397]
* control types: <br>
"ee": learn the displacement of the end effector. <br>
"joint": learn the displacement (of angles) of the 7 joints.

## Contact Info

Maintainer: Angela Wu (annwu@rice.edu) <br>
Reachable via Slack as well.







CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package franka_ros
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------


src/openai_ros/robot_gazebo_env.py
import rospy
import gym
from gym.utils import seeding
from .gazebo_connection import GazeboConnection
from .controllers_connection import ControllersConnection
#https://bitbucket.org/theconstructcore/theconstruct_msgs/src/master/msg/RLExperimentInfo.msg
from openai_ros.msg import RLExperimentInfo

# https://github.com/openai/gym/blob/master/gym/core.py
class RobotGazeboEnv(gym.Env):

    def __init__(self, robot_name_space, controllers_list, reset_controls, start_init_physics_parameters=True, reset_world_or_sim = "NO_RESET_SIM"):#reset_world_or_sim: "SIMULATION" "NO_RESET_SIM"
        """
        The superclass of all. Includes the definitions of the highest level.
        self.gazebo: connection of gazebo.
        self.controllers_object: connection to controllers.
        self.reset_controls: reset controls or not.
        self.episode_num: current episode number.
        self.cumulated_episode_reward: the episode's current cumulated reward.
        self.reward_pub: reward's publisher.
        """
        # To reset Simulations
        rospy.logdebug("START init RobotGazeboEnv")
        self.gazebo = GazeboConnection(start_init_physics_parameters,reset_world_or_sim)
        self.controllers_object = ControllersConnection(namespace = robot_name_space, controllers_list = controllers_list) ### set controller and namespace here.
        self.reset_controls = reset_controls
        self.seed()

        self.episode_num = 0
        self.cumulated_episode_reward = 0
        self.reward_pub = rospy.Publisher('/openai/reward', RLExperimentInfo, queue_size = 1)
        rospy.logdebug("END init RobotGazeboEnv")

    def seed(self, seed = None):
        """
        set a function that generate random values.
        :param seed: the seed.
        :returns: [the seed].
        """
        self.np_random, seed = seeding.np_random(seed)
        return [seed]

    def step(self, action):
        """
        execute an action in a time step, and get related data.
        :param action: the action.
        :returns: obs, reward, done, info.
        """
        rospy.logdebug("START STEP")

        self.gazebo.unpauseSim()
        self._set_action(action) #### This will change the simulator's pose.
        obs = self._get_obs()
        self.gazebo.pauseSim() # pause after getting obs so that we can get its states.
        done = self._is_done(obs)
        info = {}
        reward = self._compute_reward(obs['achieved_goal'], obs['desired_goal'], info)
        self.cumulated_episode_reward += reward
        rospy.logdebug("END STEP")

        return obs, reward, done, info

    def reset(self):
        """
        reset the environment. have to call this before step.
        :returns: the observation after reset.
        """
        rospy.logdebug("resetting RobotGazeboEnvironment")
        self._reset_sim()
        self._init_env_variables()
        self._update_episode()
        obs = self._get_obs()
        rospy.logdebug("resetting RobotGazeboEnvironment DONE")
        return obs

    def close(self):
        """
        close the environment.
        """
        rospy.logdebug("closing RobotGazeboEnvironment")
        rospy.signal_shutdown("closing RobotGazeboEnvironment")

    # Extension methods
    # ----------------------------
    def _update_episode(self): # for internal use.
        """
        publishes the cumulated reward of the episode and increases the episode number by one.
        """
        rospy.logdebug("PUBLISHING REWARD...")
        self._publish_reward_topic(
                                    self.cumulated_episode_reward,
                                    self.episode_num
                                    )
        rospy.logdebug("PUBLISHING REWARD...DONE = "+str(self.cumulated_episode_reward)+",EP = "+str(self.episode_num))
        self.episode_num += 1
        self.cumulated_episode_reward = 0

    def _publish_reward_topic(self, reward, episode_number = 1):
        """
        publish the reward to the reward topic.
        :param reward: reward of the episode.
        :param episode_number: the number of the episode.
        """
        reward_msg = RLExperimentInfo()
        reward_msg.episode_number = episode_number
        reward_msg.episode_reward = reward
        self.reward_pub.publish(reward_msg)

    def _reset_sim(self):
        """
        resets the simulation.
        :returns: if simulation successfully reset.
        """
        rospy.logdebug("RESET SIM START")
        if self.reset_controls :
            rospy.logdebug("RESET CONTROLLERS")
            # reset controllers, check all systems ready and set robot to init pose.
            self.gazebo.unpauseSim()
            self.controllers_object.reset_controllers()
            self._check_all_systems_ready()
            self._set_init_pose()
            self.gazebo.pauseSim()
            self.gazebo.resetSim()
            self.gazebo.unpauseSim()
            self.controllers_object.reset_controllers()
            self._check_all_systems_ready()
            self.gazebo.pauseSim()
        else:
            rospy.logdebug("DONT RESET CONTROLLERS")
            # check all systems ready and set robot to init pose.
            self.gazebo.unpauseSim()
            self._check_all_systems_ready()
            self._set_init_pose()
            self.gazebo.pauseSim()
            self.gazebo.resetSim()
            self.gazebo.unpauseSim()
            self._check_all_systems_ready()
            self.gazebo.pauseSim()

        rospy.logdebug("RESET SIM END")
        return True


# Abstract Methods.
# ----------------------------
    def _set_init_pose(self): # TrainEnv.
        raise NotImplementedError()

    def _check_all_systems_ready(self): # RobotEnv.
        raise NotImplementedError()

    def _get_obs(self): # TrainEnv.
        raise NotImplementedError()

    def _init_env_variables(self): # TrainEnv.
        raise NotImplementedError()

    def _set_action(self, action): # TrainEnv.
        raise NotImplementedError()

    def _is_done(self, observations): # TrainEnv.
        raise NotImplementedError()

    def _compute_reward(self, achieved_goal, desired_goal, info): # TrainEnv.
        raise NotImplementedError()

    def _env_setup(self, initial_qpos): # TrainEnv.
        raise NotImplementedError()
    
    #def _set_limit(self):
    #    raise NotImplementedError()



src/openai_ros/gazebo_connection.py
#!/usr/bin/env python

import rospy
from std_srvs.srv import Empty
from gazebo_msgs.msg import ODEPhysics
from gazebo_msgs.srv import SetPhysicsProperties, SetPhysicsPropertiesRequest
from std_msgs.msg import Float64
from geometry_msgs.msg import Vector3

class GazeboConnection():

    def __init__(self, start_init_physics_parameters, reset_world_or_sim, max_retry = 20):
        rospy.logwarn("START Init GazeboConnection")
        self._max_retry = max_retry
        self.unpause = rospy.ServiceProxy('/gazebo/unpause_physics', Empty)
        self.pause = rospy.ServiceProxy('/gazebo/pause_physics', Empty)
        self.reset_simulation_proxy = rospy.ServiceProxy('/gazebo/reset_simulation', Empty)
        self.reset_world_proxy = rospy.ServiceProxy('/gazebo/reset_world', Empty)
        # Setup the Gravity Controle system
        service_name = '/gazebo/set_physics_properties'
        rospy.logdebug("Waiting for service " + str(service_name))
        rospy.wait_for_service(service_name)
        rospy.logdebug("Service Found " + str(service_name)) ### will find service once launching panda.launch.
        self.set_physics = rospy.ServiceProxy(service_name, SetPhysicsProperties)
        self.start_init_physics_parameters = start_init_physics_parameters
        self.reset_world_or_sim = reset_world_or_sim
        self.init_values()
        # We always pause the simulation, important for legged robots learning
        self.pauseSim()
        rospy.logwarn("END Init GazeboConnection")


    def pauseSim(self):
        rospy.logdebug("PAUSING START")
        rospy.wait_for_service('/gazebo/pause_physics')
        rospy.logdebug("PAUSING service found...")
        paused_done = False
        counter = 0
        while not paused_done and not rospy.is_shutdown():
            if counter < self._max_retry:
                try:
                    rospy.logdebug("PAUSING service calling...")
                    self.pause()
                    paused_done = True
                    rospy.logdebug("PAUSING service calling...DONE")
                except rospy.ServiceException as e:
                    counter += 1
                    rospy.logerr("/gazebo/pause_physics service call failed")
            else:
                error_message = "Maximum retries done"+str(self._max_retry)+", please check Gazebo pause service"
                rospy.logerr(error_message)
                assert False, error_message

        rospy.logdebug("PAUSING FINISH")

    def unpauseSim(self):
        rospy.logdebug("UNPAUSING START")
        rospy.wait_for_service('/gazebo/unpause_physics')
        rospy.logdebug("UNPAUSING service found...")
        unpaused_done = False
        counter = 0
        while not unpaused_done and not rospy.is_shutdown():
            if counter < self._max_retry:
                try:
                    rospy.logdebug("UNPAUSING service calling...")
                    self.unpause()
                    unpaused_done = True
                    rospy.logdebug("UNPAUSING service calling...DONE")
                except rospy.ServiceException as e:
                    counter += 1
                    rospy.logerr("/gazebo/unpause_physics service call failed...Retrying "+str(counter))
            else:
                error_message = "Maximum retries done"+str(self._max_retry)+", please check Gazebo unpause service"
                rospy.logerr(error_message)
                assert False, error_message

        rospy.logdebug("UNPAUSING FiNISH")


    def resetSim(self):
        """
        This was implemented because some simulations, when reseted the simulation
        the systems that work with TF break, and because sometime we wont be able to change them
        we need to reset world that ONLY resets the object position, not the entire simulation
        systems.
        """
        if self.reset_world_or_sim == "SIMULATION":
            rospy.logdebug("SIMULATION RESET")
            self.resetSimulation()
        elif self.reset_world_or_sim == "WORLD":
            rospy.logdebug("WORLD RESET")
            self.resetWorld()
        elif self.reset_world_or_sim == "NO_RESET_SIM":
            rospy.logdebug("NO RESET SIMULATION SELECTED")
        else:
            rospy.logdebug("WRONG Reset Option:"+str(self.reset_world_or_sim))

    def resetSimulation(self):
        rospy.wait_for_service('/gazebo/reset_simulation')
        try:
            self.reset_simulation_proxy() ####### WILL HAVE TO CHANGE DEF IN /gazebo/reset_simulation?
        except rospy.ServiceException as e:
            print ("/gazebo/reset_simulation service call failed")

    def resetWorld(self):
        rospy.wait_for_service('/gazebo/reset_world')
        try:
            self.reset_world_proxy()
        except rospy.ServiceException as e:
            print ("/gazebo/reset_world service call failed")

    def init_values(self):

        self.resetSim()

        if self.start_init_physics_parameters:
            rospy.logdebug("Initialising Simulation Physics Parameters")
            self.init_physics_parameters()
        else:
            rospy.logerr("NOT Initialising Simulation Physics Parameters")

    def init_physics_parameters(self): ####  MAY BE BAD. FOLLOW FETCH'S CONFIG.
        """
        We initialise the physics parameters of the simulation, like gravity,
        friction coeficients and so on.
        """
        self._time_step = Float64(0.002) #### 0.001 -> 0.002, as seen in panda-gym pybullet.py
        self._max_update_rate = Float64(1000.0)

        self._gravity = Vector3() # correct.
        self._gravity.x = 0.0
        self._gravity.y = 0.0
        self._gravity.z = -9.81

        self._ode_config = ODEPhysics()
        self._ode_config.auto_disable_bodies = False
        self._ode_config.sor_pgs_precon_iters = 0
        self._ode_config.sor_pgs_iters = 50
        self._ode_config.sor_pgs_w = 1.3
        self._ode_config.sor_pgs_rms_error_tol = 0.0
        self._ode_config.contact_surface_layer = 0.001
        self._ode_config.contact_max_correcting_vel = 0.0
        self._ode_config.cfm = 0.0 ### first to tune?
        self._ode_config.erp = 0.2
        self._ode_config.max_contacts = 20

        self.update_gravity_call()


    def update_gravity_call(self):

        self.pauseSim()

        set_physics_request = SetPhysicsPropertiesRequest()
        set_physics_request.time_step = self._time_step.data
        set_physics_request.max_update_rate = self._max_update_rate.data
        set_physics_request.gravity = self._gravity
        set_physics_request.ode_config = self._ode_config

        rospy.logdebug(str(set_physics_request.gravity))

        result = self.set_physics(set_physics_request)
        rospy.logdebug("Gravity Update Result==" + str(result.success) + ",message==" + str(result.status_message))

        self.unpauseSim()

    def change_gravity(self, x, y, z):
        self._gravity.x = x
        self._gravity.y = y
        self._gravity.z = z

        self.update_gravity_call()

openai_ros/src/openai_ros/task_envs/fetch_reach/fetch_reach.py
from gym import utils

import rospy
from gym import spaces
from openai_ros.robot_envs import fetch_env_v2
#from my_fetch_train import fetch_env
#import fetch_env
from gym.envs.registration import register
import numpy as np
from sensor_msgs.msg import JointState
from fetch_train.srv import EePose, EePoseRequest, EeRpy, EeRpyRequest, EeTraj, EeTrajRequest, JointTraj, JointTrajRequest


register(
        id='FetchReach-v0',
        entry_point='openai_ros.task_envs.fetch_reach.fetch_reach:FetchReachEnv',
        max_episode_steps=1000,
    )


class FetchReachEnv(fetch_env_v2.FetchEnv, utils.EzPickle):
    def __init__(self):
        
        print ("Entered Reach Env")
        
        self.get_params()
        
        fetch_env_v2.FetchEnv.__init__(self)
        utils.EzPickle.__init__(self)
        
        print ("Call env setup")
        self._env_setup(initial_qpos=self.init_pos)
        
        print ("Call get_obs")
        obs = self._get_obs()
        
        self.action_space = spaces.Box(-1., 1., shape=(self.n_actions,), dtype='float32')
        self.observation_space = spaces.Dict(dict(
            desired_goal=spaces.Box(-np.inf, np.inf, shape=obs['achieved_goal'].shape, dtype='float32'),
            achieved_goal=spaces.Box(-np.inf, np.inf, shape=obs['achieved_goal'].shape, dtype='float32'),
            observation=spaces.Box(-np.inf, np.inf, shape=obs['observation'].shape, dtype='float32'),
        ))

        
    def get_params(self):
        #get configuration parameters
        """
        self.n_actions = rospy.get_param('/fetch/n_actions')
        self.has_object = rospy.get_param('/fetch/has_object')
        self.block_gripper = rospy.get_param('/fetch/block_gripper')
        self.n_substeps = rospy.get_param('/fetch/n_substeps')
        self.gripper_extra_height = rospy.get_param('/fetch/gripper_extra_height')
        self.target_in_the_air = rospy.get_param('/fetch/target_in_the_air')
        self.target_offset = rospy.get_param('/fetch/target_offset')
        self.obj_range = rospy.get_param('/fetch/obj_range')
        self.target_range = rospy.get_param('/fetch/target_range')
        self.distance_threshold = rospy.get_param('/fetch/distance_threshold')
        self.init_pos = rospy.get_param('/fetch/init_pos')
        self.reward_type = rospy.get_param('/fetch/reward_type')
        """
        self.n_actions = 4
        self.has_object = False
        self.block_gripper = True
        self.n_substeps = 20
        self.gripper_extra_height = 0.2
        self.target_in_the_air = True
        self.target_offset = 0.0
        self.obj_range = 0.15
        self.target_range = 0.15
        self.distance_threshold = 0.05
        self.reward_type = "sparse"
        self.init_pos = {
            'joint1': 0.0,
            'joint2': 0.0,
            'joint3': 0.0,
            'joint4': -1.5,
            'joint5': 0.0,
            'joint6': 1.5,
            'joint7': 0.0,
        }
        
    def _set_action(self, action):
        
        # Take action
        assert action.shape == (4,)
        action = action.copy()  # ensure that we don't change the action outside of this scope
        pos_ctrl, gripper_ctrl = action[:3], action[3]

        #pos_ctrl *= 0.05  # limit maximum change in position
        rot_ctrl = [1., 0., 1., 0.]  # fixed rotation of the end effector, expressed as a quaternion
        gripper_ctrl = np.array([gripper_ctrl, gripper_ctrl])
        assert gripper_ctrl.shape == (2,)
        if self.block_gripper:
            gripper_ctrl = np.zeros_like(gripper_ctrl)
        action = np.concatenate([pos_ctrl, rot_ctrl, gripper_ctrl])
            
            
        # Apply action to simulation.
        self.set_trajectory_ee(action)

    def _get_obs(self):
        
        grip_pos = self.get_ee_pose()
        grip_pos_array = np.array([grip_pos.pose.position.x, grip_pos.pose.position.y, grip_pos.pose.position.z])
        #dt = self.sim.nsubsteps * self.sim.model.opt.timestep #What is this??
        #grip_velp = self.sim.data.get_site_xvelp('robot0:grip') * dt
        grip_rpy = self.get_ee_rpy()
        #print grip_rpy
        grip_velp = np.array([grip_rpy.y, grip_rpy.y])
        robot_qpos, robot_qvel = self.robot_get_obs(self.joints)
        if self.has_object:
            object_pos = self.sim.data.get_site_xpos('object0')
            # rotations
            object_rot = rotations.mat2euler(self.sim.data.get_site_xmat('object0'))
            # velocities
            object_velp = self.sim.data.get_site_xvelp('object0') * dt
            object_velr = self.sim.data.get_site_xvelr('object0') * dt
            # gripper state
            object_rel_pos = object_pos - grip_pos
            object_velp -= grip_velp
        else:
            object_pos = object_rot = object_velp = object_velr = object_rel_pos = np.zeros(0)
            
        gripper_state = robot_qpos[-2:]
        gripper_vel = robot_qvel[-2:] #* dt  # change to a scalar if the gripper is made symmetric
        """
        if not self.has_object:
            achieved_goal = grip_pos_array.copy()
        else:
            achieved_goal = np.squeeze(object_pos.copy())
        """    
        achieved_goal = self._sample_achieved_goal(grip_pos_array, object_pos)
            
        obs = np.concatenate([
            grip_pos_array, object_pos.ravel(), object_rel_pos.ravel(), gripper_state, object_rot.ravel(),
            object_velp.ravel(), object_velr.ravel(), gripper_vel,
        ])

        return {
            'observation': obs.copy(),
            'achieved_goal': achieved_goal.copy(),
            'desired_goal': self.goal.copy(),
        }
        
    def _is_done(self, observations):
        
        d = self.goal_distance(observations['achieved_goal'], self.goal)
        
        return (d < self.distance_threshold).astype(np.float32)
        
    def _compute_reward(self, observations, done):

        d = self.goal_distance(observations['achieved_goal'], self.goal)
        if self.reward_type == 'sparse':
            return -(d > self.distance_threshold).astype(np.float32)
        else:
            return -d
        
    def _init_env_variables(self):
        """
        Inits variables needed to be initialized each time we reset at the start
        of an episode.
        :return:
        """
        pass
    
    def _set_init_pose(self):
        """Sets the Robot in its init pose
        """
        self.gazebo.unpauseSim()
        self.set_trajectory_joints(self.init_pos)

        return True
        
    def goal_distance(self, goal_a, goal_b):
        assert goal_a.shape == goal_b.shape
        return np.linalg.norm(goal_a - goal_b, axis=-1)
        

    def _sample_goal(self):
        if self.has_object:
            goal = self.initial_gripper_xpos[:3] + self.np_random.uniform(-self.target_range, self.target_range, size=3)
            goal += self.target_offset
            goal[2] = self.height_offset
            if self.target_in_the_air and self.np_random.uniform() < 0.5:
                goal[2] += self.np_random.uniform(0, 0.45)
        else:
            goal = self.initial_gripper_xpos[:3] + self.np_random.uniform(-0.15, 0.15, size=3)
        
        #return goal.copy()
        return goal
        
    def _sample_achieved_goal(self, grip_pos_array, object_pos):
        if not self.has_object:
            achieved_goal = grip_pos_array.copy()
        else:
            achieved_goal = np.squeeze(object_pos.copy())
        
        #return achieved_goal.copy()
        return achieved_goal
        
    def _env_setup(self, initial_qpos):
        print ("Init Pos:")
        print (initial_qpos)
        #for name, value in initial_qpos.items():
        self.gazebo.unpauseSim()
        self.set_trajectory_joints(initial_qpos)
            #self.execute_trajectory()
        #utils.reset_mocap_welds(self.sim)
        #self.sim.forward()

        # Move end effector into position.
        gripper_target = np.array([0.498, 0.005, 0.431 + self.gripper_extra_height])# + self.sim.data.get_site_xpos('robot0:grip')
        gripper_rotation = np.array([1., 0., 1., 0.])
        #self.sim.data.set_mocap_pos('robot0:mocap', gripper_target)
        #self.sim.data.set_mocap_quat('robot0:mocap', gripper_rotation)
        action = np.concatenate([gripper_target, gripper_rotation])
        self.set_trajectory_ee(action)
        #self.execute_trajectory()
        #for _ in range(10):
            #self.sim.step()
            #self.step()

        # Extract information for sampling goals.
        #self.initial_gripper_xpos = self.sim.data.get_site_xpos('robot0:grip').copy()
        gripper_pos = self.get_ee_pose()
        gripper_pose_array = np.array([gripper_pos.pose.position.x, gripper_pos.pose.position.y, gripper_pos.pose.position.z])
        self.initial_gripper_xpos = gripper_pose_array.copy()
        if self.has_object:
            self.height_offset = self.sim.data.get_site_xpos('object0')[2]
            
        self.goal = self._sample_goal()
        self._get_obs()
        
    def robot_get_obs(self, data):
        
        """
        Returns all joint positions and velocities associated with a robot.
        """
    
        if data.position is not None and data.name:
            #names = [n for n in data.name if n.startswith('robot')]
            names = [n for n in data.name]
            i = 0
            r = 0
            for name in names:
                r += 1
                
            return (
                np.array([data.position[i] for i in range(r)]),
                np.array([data.velocity[i] for i in range(r)]),
            )
        return np.zeros(0), np.zeros(0)


src/openai_ros/task_envs/reach/reach.py
from gym import spaces, utils
from gym.envs.registration import register
import rospy
from openai_ros.robot_envs import panda_env
import numpy as np
import time

timestep_limit_per_episode = 50

# register environment in gym.
register(
        id = 'PandaReach-v2',
        entry_point = 'openai_ros.task_envs.reach.reach:ReachEnv',
        max_episode_steps = timestep_limit_per_episode
    )

class ReachEnv(panda_env.PandaEnv, utils.EzPickle):
    """
    This class provides all the context for the reach task we want Panda to learn. 
    self.control_type: what kind of control to learn. "ee" or "joint". affect self.n_actions.
    self.n_actions: number of actions.
    self.has_object: if there's object in the environment.
    self.block_gripper: if we want to ignore gripper.
    self.distance_threshold: threshold distance to the goal.
    self.reward_type: the type of reward.
    self.init_pos: initial pose.
    self.n_substeps: number of substeps.
    self.gripper_extra_height: height of the gripper.
    self.target_in_the_air: if the target is in the air.
    self.target_offset: offset to the target.
    self.obj_range: the range of the object.
    self.target_range: the max distance from start to target.
    self.action_space: the action space.
    self.observation_space: the observation space.
    self.initial_gripper_pos: initial gripper position.
    """
    def __init__(self, control_type="ee"): ## default control_type is ee.
        rospy.logdebug("entered Reach Env")
        # set params.
        self.control_type = control_type
        self.get_params()

        panda_env.PandaEnv.__init__(self)
        utils.EzPickle.__init__(self)

        rospy.logdebug("reach env - setup")
        self._env_setup(self.init_pos)
 
        rospy.logdebug("reach env - check obs")
        obs = self._get_obs()
        rospy.logdebug("current obs:\n")
        rospy.logdebug(obs)

        self.action_space = spaces.Box(-1., 1., shape=(self.n_actions,), dtype='float32')
        self.observation_space = spaces.Dict(
            dict(
                observation=spaces.Box(-10., 10., shape=(6,), dtype=np.float32),
                desired_goal=spaces.Box(-10., 10., shape=obs['achieved_goal'].shape, dtype=np.float32),
                achieved_goal=spaces.Box(-10., 10., shape=obs['desired_goal'].shape, dtype=np.float32),
            )
        )
        #self._set_limit()
        rospy.logdebug("Reach Env init DONE")


    def get_params(self):
        """
        get configuration parameters.
        """
        if self.control_type == "ee":
            self.n_actions = 3 # ee position
        else:
            self.n_actions = 7 # joint angle.
        self.has_object = False
        self.block_gripper = True
        self.distance_threshold = 0.05
        self.reward_type = "sparse"
        self.init_pos = { # 90 degree bend in the elbow. from OLD panda.launch in franka_gazebo.
            'panda_joint1': 0.0,
            'panda_joint2': 0.0,
            'panda_joint3': 0.0,
            'panda_joint4': -1.57079632679,
            'panda_joint5': 0.0,
            'panda_joint6': 1.57079632679,
            'panda_joint7': 0.785398163397,
        }
        self.n_substeps = 20
        self.gripper_extra_height = 0.2
        self.target_in_the_air = True
        self.target_offset = 0.0
        self.obj_range = 0.15
        # self.target_range = 0.15
        self.goal_range_low = np.array([-0.15, -0.15, 0])
        self.goal_range_high = np.array([0.15, 0.15, 0.3])

    def robot_get_obs(self, joints):
        """
        returns current joint positions and joint velocities of Panda.
        :param joints: a list of joints.
        :returns: current joint positions and joint velocities.
        """
        if joints.position is not None and joints.name:
            names = [n for n in joints.name]
            i = 0
            r = 0
            # count the number of joints.
            for name in names:
                r += 1
                
            return (
                np.array([joints.position[i] for i in range(r)]),
                np.array([joints.velocity[i] for i in range(r)]),
            )
        return np.zeros(0), np.zeros(0)

    def compute_reward(self, achieved_goal, desired_goal, info):
        return self._compute_reward(achieved_goal, desired_goal, info)

    # RobotGazeboEnv's virtual methods.
    # --------------------------------
    def _set_init_pose(self):
        """
        sets Panda in its init pose.
        :returns: if the plan succeed.
        """
        self.gazebo.unpauseSim()
        result = self.set_trajectory_joints(self.init_pos)

        return result

    def _init_env_variables(self):
        """
        init environment variables.
        """
        self.n_actions = 3
        self.has_object = False
        self.block_gripper = True
        self.distance_threshold = 0.05
        self.reward_type = "sparse"
        if self.control_type == "ee":
            self.n_actions = 3 # ee position
        else:
            self.n_actions = 7 # joint angle.
        self.init_pos = { # 90 degree bend in the elbow. from OLD panda.launch in franka_gazebo.
            'panda_joint1': 0.0,
            'panda_joint2': 0.0,
            'panda_joint3': 0.0,
            'panda_joint4': -1.57079632679,
            'panda_joint5': 0.0,
            'panda_joint6': 1.57079632679,
            'panda_joint7': 0.785398163397,
        }
        self.n_substeps = 20
        self.gripper_extra_height = 0.2
        self.target_in_the_air = True
        self.target_offset = 0.0
        self.obj_range = 0.15
        #self.target_range = 0.15
        self.goal_range_low = np.array([-0.15, -0.15, 0])
        self.goal_range_high = np.array([0.15, 0.15, 0.3])

    def _set_action(self, action):
        """
        move the robot based on the action variable given.
        reference: ee_displacement_to_target_arm_angles in panda.py.
        :param action: the action.
        """
        if self.control_type == "ee":
            assert action.shape == (3,)
            action = action.copy()  # ensure action don't change
            action = np.clip(action, self.action_space.low, self.action_space.high)
            ee_displacement = action[:3] * 0.05
            current_obs = self._get_obs()
            next_obs = current_obs['observation'][:3] + ee_displacement
            next_obs[2] = max(0, next_obs[2])
            rot_ctrl = [1., 0., 0., 0.] ### PLACEHOLDER FOR ORIENTATION OF EE.
            action = np.concatenate([next_obs, rot_ctrl])
            self.set_trajectory_ee(action)  
        else: # "joint"
            assert action.shape == (7,)
            action = action.copy()  # ensure action don't change
            action = np.clip(action, self.action_space.low, self.action_space.high)
            joint_displacement = action * 0.05
            current_joint, _ = self.robot_get_obs(self.joints)
            next_joint = current_joint[:7] + joint_displacement
            # make joint 
            joint_degree = {
            'panda_joint1': float(next_joint[0]),
            'panda_joint2': float(next_joint[1]),
            'panda_joint3': float(next_joint[2]),
            'panda_joint4': float(next_joint[3]),
            'panda_joint5': float(next_joint[4]),
            'panda_joint6': float(next_joint[5]),
            'panda_joint7': float(next_joint[6]),
            }
            self.set_trajectory_joints(joint_degree)
    
     
    def _get_obs(self):
        """
        get the observation of the state now.
        :returns: observation (ee position and velocity), achieved goal and desired goal.
        """
        grip_pos = self.get_ee_pose()
        grip_pos_array = np.array([grip_pos.pose.position.x, grip_pos.pose.position.y, grip_pos.pose.position.z])
        grip_vel_array = np.array([0., 0., 0.])
        achieved_goal = self._sample_achieved_goal(grip_pos_array)
        obs = np.concatenate([
            grip_pos_array, grip_vel_array,
        ])

        return {
            'observation': obs.copy(),
            'achieved_goal': achieved_goal.copy(),
            'desired_goal': self.goal.copy(),
        }

    def _is_done(self, observations):
        """
        check if the episode and is done.
        :param observations: the observations of the state.
        :returns: if the episode is done.
        """
        if np.linalg.norm(observations['achieved_goal'] - observations['desired_goal'], axis = -1) < self.distance_threshold:
            return True
        else:
            return False

    def _compute_reward(self, achieved_goal, desired_goal, info):
        """
        return the reward based on the observations given.
        :param observations: the observations of the state.
        :param done: if the episode is done.
        :returns: the reward.
        """
        distance = np.linalg.norm(achieved_goal - desired_goal, axis = -1)
        if self.reward_type == "sparse":
            return -np.array(distance > self.distance_threshold, dtype = np.float64)
        else:
            return -distance
    
    #def _set_limit(self):
    #    self.position_low = self.initial_gripper_pos + [-self.distance_threshold] * len(self.initial_gripper_pos)
    #    self.position_high = self.initial_gripper_pos + [self.distance_threshold] * len(self.initial_gripper_pos)
    #    self.joint_low = []
    #    self.joint_high = []
    # --------------------------------

    # internal methods for ReachEnv.
    # -----------------------------
    def _sample_achieved_goal(self, grip_pos_array):
        """
        return the achieved goal.
        :param grip_pos_array: the position array of the gripper.
        :returns: the achieved goal.
        """
        achieved_goal = grip_pos_array.copy()
        return achieved_goal

    def _sample_goal(self):
        """
        sample a goal according to self.target_range.
        :returns: the goal.
        """
        #goal = self.initial_gripper_pos[:3] + self.np_random.uniform(-self.target_range, self.target_range, size = 3)
        goal = self.initial_gripper_pos[:3] + self.np_random.uniform(self.goal_range_low, self.goal_range_high) 
        
        return goal

    def _env_setup(self, initial_pos):
        """
        setup the init pose of the robot and the gripper, and set a goal.
        :param initial_pos: initial pose of the joints.
        """
        self.gazebo.unpauseSim()
        self.set_trajectory_joints(initial_pos)

        # get the position of the end effector.
        gripper_pos = self.get_ee_pose()
        gripper_pose_array = np.array([gripper_pos.pose.position.x, gripper_pos.pose.position.y, gripper_pos.pose.position.z])
        self.initial_gripper_pos = gripper_pose_array.copy()
            
        self.goal = self._sample_goal()
    # -----------------------------